llvm/include/llvm/ADT/ArrayRef.h

```diff
@@ -149,6 +149,20 @@ namespace llvm {
                  * = nullptr)
         : Data(Vec.data()), Length(Vec.size()) {}
 
+    /// Construct an ArrayRef<T> from iterator_range<U*>. This uses SFINAE
+    /// to ensure that this is only used for iterator ranges of random access
+    /// iterators that can be converted.
+    template <typename U,
+              typename = std::enable_if<
+                  std::is_base_of<
+                      std::random_access_iterator_tag,
+                      typename std::iterator_traits<
+                          decltype(std::declval<const iterator_range<U *> &>()
+                                       .begin())>::iterator_category>::value &&
```

Comment:
I don't think we need this declval, but I could be wrong: does `iterator_traits<decltype(iterator<U *>)>::iterator_category` not work? Why use iterator_range::begin(), when we already know what the type of the begin is?